### PR TITLE
Add graph controls to change graph dimensions and transfer dimensions to...

### PR DESCRIPTION
**Pull request by psytra**
... other graphs in a collection (#188)

Looks pretty good to me...

### CODE REVIEW

**Comment by vbatoufflet** [comment]
Sorry, I still didn't had the time to test your PR.

I'll let you know as soon as it will be done :wink:

**Comment by vbatoufflet** [comment]
Hi Matthias,

I started to review your PR and I have a few remarks.

I am wondering if graph dimensions shouldn't be handled at collection level. Your PR adds many controls in graph's toolbar and I think it makes it more complex. Handling it in collection definitions will prevent dimensions being lost when reloading the page too.

I don't think it is necessary to allow width change. But, just for the record, we are planning to have a grid system in a future release to allow a dashboard-like display with possibly different heights and widths for every elements.

WDYT?

**Comment by psytra** [comment]
**Graph Dimensions at Collection Level**
Well, the control element bar gets bloated, yes. OTOH I think analyzing a graph shouldn't require you to edit graph preferences. So Zoom + size should be dynamic.

**Reloading**
Yes, that's an issue. But not more or less of an issue than what we have with time ranges or zooming right now. 
The problem also exists when you disable series in the graph. They are also reset.

**Width change**
First I implemented width adjustment just for fun. Right now I think this is a life-changer, as you can go really wide and have much bigger time ranges with the same precision.

**Conclusion**
I have to admit, that I expected your opinion on the panel bloat :-) But I didn't have a better idea at hand. I could imagine a mouse based resize in the graph corner. For me personally, width is a requirement, now, when I think of a perfect front end.

**Comment by falzm** [comment]
Hi @psytra 

Altough I'm OK with charts height resizing, I too am not very fond of changing width...

**Comment by psytra** [comment]
Hi @falzm, @vbatoufflet 
could you elaborate on the reasons why you prefer not having the feature? I'm OK with maintaining my own patchsets, but I'm very interested in your intentions.

Is it because you think it is not necessary for performance analysis, or not OK from UI design perspective.

I'm currently thinking about a workaround to view larger time ranges with fixed width. But that might get even crazier.. :-)

**Comment by vbatoufflet** [comment]
Hi Matthias,

My main concern is form UI point of view. Lowering the width of the graph leaves a blank area on the right and raising it makes the page scrolling horizontally. For me it seems odd :smile: 

Regarding the height, I'm totally OK with it. But I would rather a have grip handle than new buttons in the graph toolbar.

**Comment by psytra** [comment]
Well, I have to agree... partially :-)  
I still think, analyzing/interpreting graphs with timeranges >= 14days would benefit a lot from additional width. OTOH the user experience is not the best with the PR as it is right now (panel moves, as it is aligned at the right).

Yes, reducing width is odd. Scrolling is not IMO. At least with my 1600 or 1920 resolution.
I'll see if I can introduce a grip handle.

**Comment by vbatoufflet** [comment]
To explain further my reluctance about the width change, if we only raise the width of the first graph for example, it seems OK (_Figure A_). But as soon as we scroll it will be the others graphs that will have a blank area on the right too (_Figure B_). The same goes when reducing below the page width.

graphs-scroll

What about figuring a way to have **all graphs** widths changed at once, not allowing the UI to lower the width bellow the original viewport one?

**Comment by psytra** [comment]
@vbatoufflet thanks for the explanation and the graphics..
From my understanding, we could find a consensus when 
# all graphs get resized instead of individual ones 
# the graph controls follow when scrolling horizontally (always visible)
# resizing below default is prevented

With these changes I think it would be best to have some controls at the top of a collection/graph page?

**Comment by vbatoufflet** [comment]
Hi Matthias,

Sorry, I completely forgot to reply to you.

Indeed, I guess it could be done that way. I can't figure what exactly the final result will look like, but this consensus seems legit.

**Comment by psytra** [comment]
Hi Vincent,
no problem.. what do you think about that one? It's just a rough idea, ignore the missing style sheets.

dimensions

Initially I had the idea to measure width in days or hours, so that users can pick how many days they like to see in the graph. I think having a factor is practically the same, without messing around with time units, which would be dependent on the selected timerange.

**Comment by vbatoufflet** [comment]
Hi Matthias,

I definitively prefer this approach. Which boundary have chose, _10x_ ?

I would prefer a factor too. Managing changes with graphs having different ranges would be a mess :smile:

**Comment by psytra** [comment]
Yes I've picked 1x-10x with steps of 1 for width, and 1x-5x with steps of 0.1 for height.
It was a first shot, but I think these settings are sane in the most cases.

**Comment by psytra** [comment]
Hi all,
just want to let you know, that i've made some progress:

- [x] all graphs get resized instead of individual ones
- [x] the graph controls follow when scrolling horizontally (always visible)
- [x] resizing below default is prevented
but a new one is missing
- [] let title and legend table follow scrolling

and right now, I don't have a clue why the last one doesn't work, right now.

That's it for now.
Matthias

**Comment by vbatoufflet** [comment]
Hi Matthias,

Thanks for the update :+1: 

I will try to take few minutes today or tomorrow to take a look at the changes you made on your branch.

**Comment by psytra** [comment]
@vbatoufflet nothing there, yet. I think I'll first try to make the title/legend/toggle-buttons scrolling, and after that clean everything up and make a commit. So you don't have to review sth. half-baked.

**Comment by psytra** [comment]
So I pushed a preliminary version with the three checked issues addressed.
I've added a comment to the scrollHandler function in graph.js, that contains a jquery call, which can be executed in the browser console, to shift the legend. I'm still not 100% sure how to make it work nicely in the scroll handler as the legend it is created using Highcharts' renderer. 
 
While writing this message, I made another commit to make the titles scroll. That should work now.
So only the toggle legend button and the legend itself needs to be fixed.

**Comment by psytra** [comment]
So now there is everything in there, I think. Now, we can probably discuss what could be improved, or if it can get acceptable at all.

Don't know why the travis builds fail, though. make test runs fine over here. Sorry for the many commits...

Matthias

**Comment by vbatoufflet** [comment]
Hi Matthias,

Thanks for commiting your work. I will take a closer look at it.

**Comment by vbatoufflet** [comment]
Hi Matthias,

I took a look at your PR and I like this version way better.

Except for some minor UI improvements (lack of margin and/or padding on slides and at the right of the page when width is raised) it seems very promising :+1: 

I will take a look at the code and make you some comments on it.

Thanks for your work on this.

File: cmd/facette/template/browse/graph.html [hunk 1]
```diff
@@ -43,6 +43,23 @@
...
 							</div>
 						</li>
+						<li>
+							<a class="icon icon-size" href="#set-global-size" title="Set graph dimensions"></a>
+							<div id="widthmenu" class="menu">
+								<div class="menucntr">
+									<div class="menumesg">Set Graph Width: <em id="widthlabel">1x</em></div>
+									<div class="menuitem">
+										<input type="range" name="width" class="size-slider" step="1"
+										    min="1" max="10" value="1" oninput="$('#widthlabel').html(this.value + 'x');"/>
```
Comment (vbatoufflet) [inline cmd/facette/template/browse/graph.html]: Please avoid using Javascript directly in the HTML source. Try to bind the events directly from the js files.

**Comment by psytra** [comment]
Do you have an ideas how the margins/paddings could get fxed? Another box around the scrollarea?
I'm not 100% sure If I got it.

**Comment by vbatoufflet** [comment]
I will take a closer look and see if I can help you with this.

**Comment by vbatoufflet** [comment]
No longer applicable as Facette 0.4 has been rewritten and now uses [D3.js](https://d3js.org/).

**Pull request closed without merging**